### ISSUE TITLE
SSL/TLS requirements

### ISSUE THREAD

**Issue by szepeviktor** [lycheeverse/lychee#1232]
Hello I use these wget options to download stuff.
```bash
wget --secure-protocol=TLSv1_3 --max-redirect=1 --retry-on-host-error --retry-connrefused --tries=3
```
I think all of these can be told to lychee except minimum TLS version.

Do you plan to support that?

**Comment by mre** [comment]
Yeah, good idea.
Some of it is already supported.

We have

- `--max-redirects`, which defaults to 5
- `--max-retries`, which defaults to 3

We are missing

- `--retry-on-host-error`: I'm undecided about that one. The docs say: `Consider host errors, such as “Temporary failure in name resolution”, as non-fatal, transient errors”`. From my experience, DNS errors are usually not resolved within a couple seconds in between retries, so maybe the use for that is limited? 
- `--retry-connrefused`: Same here; connection issues are rarely transient, from what I saw.

For reference, our retry handling is at https://github.com/lycheeverse/lychee/blob/master/lychee-lib/src/retry.rs.
We can discuss the missing options.

**Comment by mre** [comment]
Interesting use-case

**Comment by mre** [comment]
I'm not against adding support for that, but I will not add it myself. I'd appreciate a pull request, though.
The way I currently think about it would be to add `--min-tls` arg, which supports constants such as `--min-tls=TLSv1_3`. We could reuse the wget naming conventions for the variants if it makes sense. Open for suggestions.

**Comment by szepeviktor** [comment]
All right.
For me a Hello World is impossible in Rust as I do not know the syntax.

**Comment by HU90m** [comment]
The implementation isn't too complicated.

There is, however, one hicup: [`reqwest::ClientBuilder::min_tls_version`](https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html#method.min_tls_version) states the following.

> A value of `tls::Version::TLS_1_3` will cause an error with the `native-tls`/`default-tls` backend. This does not mean the version isn’t supported, just that it can’t be set as a minimum due to technical limitations.

Looking at https://github.com/seanmonstar/reqwest/pull/1315, it appears this is because `native-tls` doesn't allow one to specify the use of TLSv1_3, and so it can't be set as the minimum version.

**Comment by mre** [comment]
Good progress. I don't know what we could do with TLS 1.3. Maybe we print an error with nativetls and support it with rustls?

**Comment by mre** [comment]
@HU90m, any chance of converting your WIP branch into a pull request? I'd be interested in merging the feature into lychee.

**Comment by HU90m** [comment]
Certainly! It's now on my todo list. I have a busy month ahead, but will aim to find time to open a PR before feb.

**Comment by mre** [comment]
You're awesome. 🌟

**Comment by HU90m** [comment]
Really sorry [this](https://github.com/lycheeverse/lychee/pull/1655) took me so long to get around to doing. I still need to fix tests and add documentation then I'll mark it for review.

**Comment by mre** [comment]
> I'm not against adding support for that, but I will not add it myself. I'd appreciate a pull request, though. The way I currently think about it would be to add `--min-tls` arg, which supports constants such as `--min-tls=TLSv1_3`. We could reuse the wget naming conventions for the variants if it makes sense. Open for suggestions.

This syntax just landed in `master` thanks to @HU90m. You can now configure the minimum accepted TLS version, e.g.:

```sh
lychee --min-tls=TLSv1_2 https://example.com
```
